### DISCUSSION TITLE
Local Variables?

### DISCUSSION THREAD

**Discussion by JaredDavis22** [tomj74/chunk-templates#39] [Q&A] [answered]
Hi,

Thanks for writing Chunk! 

Is it possible to set a local variable in a loop?

Is it possible to break out of a loop? I can use a local variable set to skip all following iterations.

I process a list of items. Once I see a flag in the items I want to do something different for the following items until I see another flag.

Something like this:
```
for x in items
  if (x.hasFlag) {
    set local variable treat=1
  }
  if (x.hasFlagTwo) {
    set local variable treat=0
  }
  if (treat == 1) {
    format x this way
  else
    format x that way
  }
end for x
```

Without the local variable I think I have to pre-process the items prior to calling Chunk to put the "treat" flag into the object. That takes away some of the flexibility of having the logic in the chunk template.

Regards,

Jared

**Comment by JaredDavis22** [comment]
This is ugly. Will it always work?

I created a filter to set a variable.

Calling code:

 chunk.set("UglyUgly","Ugly"); // used for set filter hack
 theme.registerFilter( new ChunkSetHackUgly() );

Then in the template I have this:

 What is xx ? {$xx}
 {$UglyUgly|set(xx,ItWorked)} 
 What is xx ?   {$xx}


Output is this:


 What is xx ? {$xx}

 What is xx ? ItWorked

I ended up with an extra blank line in error. Won't matter for my application.


And java filter class is this:

// ugly way to use filters to "simulate" a set command
// requires a $tag that exists. Suggest $UglyUgly=Ugly

public class ChunkSetHackUgly   extends BasicFilter implements ChunkFilter {

    public String transformText(Chunk chunk, String text, FilterArgs args)
    {
        if (text == null) return null;

        String[] parsedArgs = args.getUnresolvedArgs();
        if (parsedArgs == null) return "FAIL-No Arguments to SET. need (name,value) 001";

        String name = parsedArgs[0];
        if (name == null) return "FAIL-No Arguments to SET. need (name,value) 002";
        String value = parsedArgs[1]; // index out of range exception?
        if (value == null) return "FAIL-No Arguments to SET. need (name,value) 003";

        chunk.set(name,value);

        return ""; // return blank string
    }

    public String getFilterName()
    {
        return "set";
    }
}

**Comment by tomj74** [comment, marked as answer]
Well, you're right to suspect that order of execution in the rendering
process is not 100% guaranteed to be left to right. 

Loops are rendered in
the order you would expect, though, so referencing something that was "set"
in the previous loop iteration as you are hoping to do is pretty safe.


Brilliant workaround, btw, I would not have come up w/this - it does sort
of violate the paradigm a bit but it's very practical. 

On Mon, 20 Dec
2021 07:10:48 -0800, JaredDavis22  wrote:  

This is ugly. Will it always
work? 

I created a filter to set a variable. 

Calling code:


chunk.set("UglyUgly","Ugly"); // used for set filter hack

theme.registerFilter( new ChunkSetHackUgly() ); 

Then in the template I
have this: 

What is xx ? {$xx}
 {$UglyUgly|set(xx,ItWorked)}
 What is xx ?
{$xx} 

Output is this: 

What is xx ? {$xx} 

What is xx ? ItWorked 

I
ended up with an extra blank line in error. Won't matter for my
application. 

And java filter class is this: 

// ugly way to use filters
to "simulate" a set command
 // requires a $tag that exists. Suggest
$UglyUgly=Ugly 

public class ChunkSetHackUgly extends BasicFilter
implements ChunkFilter { 
public String transformText(Chunk chunk, String
text, FilterArgs args)
{
 if (text == null) return null;

 String[]
parsedArgs = args.getUnresolvedArgs();
 if (parsedArgs == null) return
"FAIL-No Arguments to SET. need (name,value) 001";

 String name =
parsedArgs[0];
 if (name == null) return "FAIL-No Arguments to SET. need
(name,value) 002";
 String value = parsedArgs[1]; // index out of range
exception?
 if (value == null) return "FAIL-No Arguments to SET. need
(name,value) 003";

 chunk.set(name,value);

 return ""; // return blank
string
}

public String getFilterName()
{
 return "set";
}

} 

--
Reply to
this email directly, view it on GitHub [1], or unsubscribe [2].
Triage
notifications on the go with GitHub Mobile for iOS [3] or Android [4]. 
You
are receiving this because you are subscribed to this thread.Message ID:
@github.com>  

 

Links:
------
[1]
https://github.com/tomj74/chunk-templates/discussions/39#discussioncomment-1847740
[2]
https://github.com/notifications/unsubscribe-auth/AAU63F5QQTYCLPV43WQG2BTUR5BPRANCNFSM5KMCA7TA
[3]
https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675
[4]
https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub

**Comment by JaredDavis22** [comment]
Thanks for the compliment! 

I dropped the chunk.set and eliminated the extra line by doing this in the template:

What is xx ? {$xx}{$UglyUgly:|set(xx,ItWorked)}
What is xx ? {$xx}

Produces

What is xx ? {$xx}
What is xx ? ItWorked

It is practical but very ugly.

**Comment by JaredDavis22** [comment]
Hi,

I added a concat filter to continue my off label usage. Is this safe? Alternatives (besides repeating the pdfurl definition in all the lines?) The tag names are defined elsewhere and can not be easily changed.

...
  {% if ($overlay) }
         **{$baseurl|concat(TEST.pdf)|set(pdfurl)}**
...
[ttemplate.txt](https://github.com/tomj74/chunk-templates/files/7774757/ttemplate.txt)

**Reply by tomj74** [reply]
Seems fine. Exec macro could work here too but this is nice and concise.

Filter chains are always processed in order.

**Comment by JaredDavis22** [comment]
Happy holidays.

Thanks for the ExecFilter hint. That looks almost perfect for my usage.

I created a copy of class ExecFilter with a slight tweak to pass parameters

...
 // exec self-contained chunk with own private context
        Chunk miniMacro = new Chunk();
        miniMacro.append(filterBody);
        miniMacro.setOrDelete("x", text);
        **for (int p=1; p< args.length; p++) {
            miniMacro.setOrDelete("p"+p, args[p]);
        }**
        return miniMacro.toString();
...

called it filterp

then changed my long case list of Ifs to be this

{$basefile|filterp(#OVERLAYPAGEP,1)}
{.if ($AA) }  {$basefile|filterp(#OVERLAYPAGEP,2)} {/if}
{.if ($AB) }  {$basefile|filterp(#OVERLAYPAGEP,3)} {/if}
{.if ($BA) }  {$basefile|filterp(#OVERLAYPAGEP,4)} {/if}
{.if ($BB) }  {$basefile|filterp(#OVERLAYPAGEP,5)} {/if}
{.if ($BC) }  {$basefile|filterp(#OVERLAYPAGEP,6)} {/if}
{.if ($BD) }  {$basefile|filterp(#OVERLAYPAGEP,7)} {/if}

Using



{!-- $x is from filter --}
{!-- $p1 is from arguments to filterp --}
{#OVERLAYPAGEP}
    <fo:block-container position="absolute" top="0pt" left="0pt" font-size="0pt" >
      <fo:block>
          <fo:external-graphic src="{$x}#page={$p1}"/>
      </fo:block>
    </fo:block-container>
{#}

**Comment by JaredDavis22** [comment]
Here is where I ended up:

Changed "filterp" to "sendto" since it read better for my usage.

added yet another filter to include a prefix for a non empty tag.

`...`
`{$basefile|sendto(#OVERLAYPAGEP)}`
`  {$baseurl|concat(CCC.pdf)|set(pdfurl)}`
`  {$pdfurl|sendto(#OVERLAYPAGEP,1)}`
`  {.if ($AA) }   {$pdfulr|sendto(#OVERLAYPAGEP,2)} {/if}`
`  {.if ($AB) }   {$pdfulr|sendto(#OVERLAYPAGEP,3)} {/if}`
`..`

{#OVERLAYPAGEP}
    <fo:block-container position="absolute" top="0pt" left="0pt" font-size="0pt" >
      <fo:block>
          <fo:external-graphic src="{$x}{$p1:|prefix(#page=)}"/>
      </fo:block>
    </fo:block-container>
{#}

samples:

{$basefile|sendto(#OVERLAYPAGEP)} ==> ... <fo:external-graphic src="file:yada/AAA.sub.pdf"/>...
{$basefile|sendto(#OVERLAYPAGEP,5)}  ==>  ...<fo:external-graphic src="file:yada/AAA.sub.pdf#page=5"/>...

**Comment by JaredDavis22** [comment]
My ugly set code failed in this usage pattern. Is there anything I can do to influence evaluation order. I'm assuming evaluation order is the issue. I double checked the set code was triggering. I was trying to use a flag on one item to alter formatting on the next item. I tried to sprinkle some $x values in different blocks. No joy.

The %else% block never shows $logonext as non blank.

      {% loop in $items as $x %}
        {% if ($x.level == "10") %}
             {$x.text} as 10
        {% elseIf ($x.level == "0") %}
             {$x.text} as 0
        {% elseIf ($x.is_format_as_wide) %}
             {$x.text} as wide
        {% else %}
              {$x.col_one:$x.text} {$x.col_two:} {$x.uom:}
              {% if ($logonext) %}
                showlogo ! resetting logonext was removed for this test
              {% endif %}
        {% endif %}
        {% if ($x.level == "0") && ($x.is_show_logo) %}
          {$logo|set(logonext)}
        {% endif %}
      {% endloop %}
